tools/ktx/validate.cpp

```diff
@@ -0,0 +1,1577 @@
+        }
+    }
+
+    template <typename... Args>
+    void error(const IssueError& issue, Args&&... args) {
+        ++numError;
+        callback(ValidationReport{issue.type, issue.id, std::string{issue.message}, fmt::format(issue.detailsFmt, std::forward<Args>(args)...)});
+    }
+
+    template <typename... Args>
+    void fatal(const IssueFatal& issue, Args&&... args) {
+        ++numError;
+        const auto report = ValidationReport{issue.type, issue.id, std::string{issue.message}, fmt::format(issue.detailsFmt, std::forward<Args>(args)...)};
+        callback(report);
+
+        throw FatalValidationError(report);
+    }
+
+private:
+    virtual void read(std::size_t offset, void* readDst, std::size_t readSize, std::string_view name) = 0;
+    virtual KTX_error_code createKTXTexture(ktxTextureCreateFlags createFlags, ktxTexture2** newTex) = 0;
+
+private:
+    template <typename... Args>
+    void validatePaddingZeros(const void* ptr, const void* bufferEnd, std::size_t alignment, const IssueError& issue, Args&&... args) {
+        const auto* begin = static_cast<const char*>(ptr);
+        const auto* end = std::min(bufferEnd, align(ptr, alignment));
+
+        for (auto it = begin; it != end; ++it)
+            if (*it != 0)
+                error(issue, static_cast<uint8_t>(*it), std::forward<Args>(args)...);
+    }
+
+public:
+    int validate();
+
+private:
+    void validateHeader();
+    void validateIndices();
+    void calculateExpectedDFD(VkFormat format);
+    // void validateLevelIndex();
+    void validateDFD();
+    void validateKVD();
+    // void validateSGD();
+    // void validateCreateAndTranscode();
+
+    void validateDFDBasic(uint32_t blockIndex, const uint32_t* dfd, const BDFD& block, const std::vector<SampleType>& samples);
+
+    void validateKTXcubemapIncomplete(const uint8_t* data, uint32_t size);
+    void validateKTXorientation(const uint8_t* data, uint32_t size);
+    void validateKTXglFormat(const uint8_t* data, uint32_t size);
+    void validateKTXdxgiFormat(const uint8_t* data, uint32_t size);
+    void validateKTXmetalPixelFormat(const uint8_t* data, uint32_t size);
+    void validateKTXswizzle(const uint8_t* data, uint32_t size);
+    void validateKTXwriter(const uint8_t* data, uint32_t size);
+    void validateKTXwriterScParams(const uint8_t* data, uint32_t size);
+    void validateKTXastcDecodeMode(const uint8_t* data, uint32_t size);
+    void validateKTXanimData(const uint8_t* data, uint32_t size);
+};
+
+// -------------------------------------------------------------------------------------------------
+
+class ValidationContextStream : public ValidationContext {
+private:
+    FileGuard fileGuard; /// Only used if the file is owned by the validation context
+    FILE* file;
+
+public:
+    ValidationContextStream(bool warningsAsErrors, std::function<void(const ValidationReport&)> callback, FILE* file) :
```

Comment:
Why did you decide to use stdio in the new validate? I redid ktx2check a while ago to use fstream and istream. I found it a lot nicer.

Reply:
The primary reason was the native support in the libraries while the ktxStream functions (like `ktxPrintInfoForStream` and `ktxPrintKTX2InfoJSONForStream`) were not always exposed.
I will happily switch to iostreams.

Reply:
Please switch to iostreams.

That `ktxPrintInfo*ForStream` is not exposed is an oversight that happened when ktxStream was being exposed.

ktxStream was created to unify stdio stream and memory stream handling for the library's C API as only the GNU c lib supports memory streams. Much later someone suggested and  provided a PR to expose `ktxStreams`, with slight modifications, and `ktxTexture[12]_CreateFromStream, etc. The changes to ktxStream enabled one to be wrapped around other I/O paradigms. The primary goal was to wrap c++ iostreams so you can create a `ktxTexture` from the content of such a stream. See `utils/sbufstream.h` for an implementation. This is used in `ktx2check::ktxValidator::validateTranscode`.

Reply:
Done